pkg/gadgets/internal/tchandler/tracer.go

```diff
@@ -0,0 +1,291 @@
+	users map[uint32]struct{}
+}
+
+func (t *Handler) closeAttachment(a *attachment) {
+	if a.filter != nil {
+		t.tcnl.Filter().Delete(a.filter)
+	}
+	a.dispatcher.Close()
+}
+
+type Handler struct {
+	// dispatcher map is a program array map with a single element that is used by the
+	// dispatcher to perform a tail call to the gadget program.
+	dispatcherMap *ebpf.Map
+	// key: network interface name on the host side
+	// value: attachment
+	attachments map[string]*attachment
+
+	// socket to talk to netlink
+	// TODO: Currently we keep once instance of the socket for each Handler instance. Check if
+	// it makes sense to move this to the tracer to have one single instance per gadget.
+	// https://github.com/inspektor-gadget/inspektor-gadget/pull/2376#discussion_r1475472725
+	tcnl *tc.Tc
+
+	direction AttachmentDirection
+}
+
+func NewHandler(direction AttachmentDirection) (*Handler, error) {
+	var err error
+	var tcnl *tc.Tc
+
+	// We need to create the client on the host network namespace, otherwise it's not able to
+	// create the qdisc and filters.
+	err = netnsenter.NetnsEnter(1, func() error {
+		// Setup tc socket for communication with the kernel
+		tcnl, err = tc.Open(&tc.Config{})
+		if err != nil {
+			return fmt.Errorf("opening rtnetlink socket: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	t := &Handler{
+		attachments: make(map[string]*attachment),
+		tcnl:        tcnl,
+		direction:   direction,
+	}
+	defer func() {
+		if err != nil {
+			t.Close()
+		}
+	}()
+
+	// Keep in sync with tail_call map in bpf/dispatcher.bpf.c
+	dispatcherMapSpec := ebpf.MapSpec{
+		Name:       tailCallMapName,
+		Type:       ebpf.ProgramArray,
+		KeySize:    4,
+		ValueSize:  4,
+		MaxEntries: 1,
+	}
+	t.dispatcherMap, err = ebpf.NewMap(&dispatcherMapSpec)
+	if err != nil {
+		return nil, fmt.Errorf("creating tail call map: %w", err)
+	}
+	return t, nil
+}
+
+func (t *Handler) AttachProg(prog *ebpf.Program) error {
+	return t.dispatcherMap.Update(uint32(0), uint32(prog.FD()), ebpf.UpdateAny)
+}
+
+func (t *Handler) newAttachment(pid uint32, iface *net.Interface, netns uint64, direction AttachmentDirection) (_ *attachment, err error) {
+	a := &attachment{
+		users: map[uint32]struct{}{pid: {}},
+	}
+
+	var qdisc *tc.Object
+
+	defer func() {
+		if err != nil {
+			t.closeAttachment(a)
+			if qdisc != nil {
+				t.tcnl.Qdisc().Delete(qdisc)
+			}
+		}
+	}()
+
+	dispatcherSpec, err := loadDispatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	consts := map[string]interface{}{
+		"current_netns": uint32(netns),
+	}
+	if err := dispatcherSpec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("RewriteConstants while attaching to pid %d: %w", pid, err)
+	}
+
+	// We create the clsact qdisc and leak it. We can't remove it because we'll break any other
+	// application (including other ig instances) that are using it.
+	if qdisc, err = createClsActQdisc(t.tcnl, iface); err != nil {
```

Comment:
`createClsActQdisc` should return a bool saying whether it created the qdisc or it already exists in the system. Otherwise, we risk deleting the qdisc that others are using. Or any other approach, the point is to delete it only if we are sure we created it.

Reply:
Do you have a specific example where it could happen? We're only deleting the qdisc when there is an error and if the qdisc already exists `createClsActQdisc` returns nil as an error, so I think we're all fine there.

Reply:
Ah right, it can happen if the error is in LoadAndAssign() or addTCFilter(). I updated the logic to handle the case when the filter exists in tracer.go instead of tc.go

Reply:
Exactly!